Fetch translation chunks plugin

yuliyan opened the pull request:
#### Changes proposed in this Pull Request

This PR introduces the initial implementation of using chunked translations. When this feature is enabled, Calypso will switch from fetching a single file, that contains the complete set of strings, to fetching separate translation chunks bundled with each Calypso build.

In order to achieve that, it adds a few additional steps to the build process, starting with generating chunks map. Chunks map is a JSON file, that contains the paths to all JS modules keyed by the chunk they were imported to, generated by a webpack plugin.

Next step is to download `calypso-strings.pot` and translation JSON files for all languages from `https://widgets.wp.com/languages/calypso`. Previously generated chunks map is used together with `calypso-strings.pot` to get the set of strings per chunk based on the reference comments in the POT file, which is then used with the actual translations to generate individual translation chunks for each language. 

Another webpack plugin will inject additional code to [The Manifest](https://webpack.js.org/concepts/manifest/) that will expose globally a callback called whenever webpack attempts to require a JS chunk. `__requireChunkCallback__` object is attached to the window object and used to fetch translation chunks in the runtime. 

On boot, the language manifest for user's language will be downloaded. It contains the locale data and a set of all chunk ids that have translation chunks and is used to filter translation chunk request that don't exist. Then, an event handler is attached to the exposed require chunk callback to fetch translation chunks as whenever a JS chunk is being fetched in the runtime.

This is an experimental feature and therefore it won't be enabled by default. Production Docker build will also build translation chunks, but client will be still using the standard translation files, unless `use-translation-chunks` is enabled via URL flag (e.g. adding `?flags=use-translation-chunks` to the URL).

#### Testing instructions

- Boot Calypso with `ENABLE_FEATURES=use-translation-chunks npm start`
- Open http://calypso.localhost:3000/

See p4TIVU-9nl-p2

yuliyan committed:
Fetch translation chunks initial

yuliyan committed:
Use `addTranslations` instead of setLocale

yuliyan committed:
Update localeSlug manually as a temporary solution

yuliyan committed:
Rework translation chunks fetching wip

yuliyan committed:
Add generate chunks map plugin initial code

akirk commented:
Great work, this now looks a lot cleaner!

sgomes commented on client/boot/locale.js:
```diff
@@ -23,5 +24,26 @@ export const setupLocale = ( currentUser, reduxStore ) => {
+			const translationChunkName = `${ currentUser.localeSlug }-${ chunkId }`;
+
+			promises.push(
+				new Promise( resolve => {
```
Nit: wrapping in a new promise is probably not needed here. E.g. (untested):

```js
const translationChunkPath = `${ publicPath }languages/${ translationChunkName }.json`;

promises.push(
  return window
    .fetch( translationChunkPath )
    .then( response => response.json() )
    .then( data => {
      i18n.addTranslations( data );
      return; // Resolves to undefined if everything succeeds
    } )
    .catch( error => error ); // Resolves to the error if an error happens
  } )
);
```

sgomes commented on client/boot/locale.js:
```diff
@@ -23,5 +24,26 @@ export const setupLocale = ( currentUser, reduxStore ) => {
 		reduxStore.dispatch( setLocale( currentUser.localeSlug, currentUser.localeVariant ) );
 	}
 
+	if ( '__requireChunkCallback__' in window ) {
```
@Automattic/team-calypso: Do we have any guidance on adding new globals?

blowery replied:
@sgomes No real guidance, though this seems mostly OK. Maybe we should consider namespacing it, like `__a8c_requireChunkCallback__`, but that's just an abundance of caution.

yuliyan committed:
Add chunks map script

yuliyan committed:
Add CircleCI chunks map job

sgomes reviewed (commented):
Thanks for the PR, @yuliyan! I added some minor comments for now, but the idea generally looks good. I'll see if I can run some perf tests on this before a final review.

yuliyan committed:
Add trailing slash to chunks-map artifact directory

yuliyan committed:
Add chunks artifact directory

yuliyan committed:
Remove unnecessary promise wrap

akirk commented on client/webpack/fetch-translation-chunks-plugin.js:
```diff
@@ -0,0 +1,70 @@
+ */
+const PLUGIN_NAME = 'FetchTranslationChunks';
+
+class FetchTranslationChunksPlugin {
```
Given the slightly changed function of the plugin, shall we rename it? e.g. `ChunkCallbackPlugin`?

yuliyan committed:
Disable initial language fetch and fetch only translation chunks WIP

yuliyan committed:
Rename fetch-translation-chunks-plugin

yuliyan committed:
Add loaded translation chunks cache

yuliyan committed:
Expose installed chunks to be used to fetch initial translation chunks for bootstrapped chunks

yuliyan committed:
Rename chunks-map script

blowery commented:
Is there an architectural description of how this works somewhere? I'm curious about how it deals with the contents of a chunk changing over time. Every release could change the chunk makeup.

blowery commented on client/webpack/generate-chunks-map-plugin.js:
```diff
@@ -0,0 +1,39 @@
+
+			const chunksMap = chunks.reduce( ( map, chunk ) => {
+				const files = chunk.files || [];
+				const name = files.find( file => /\.js$/.test( file ) ) || files[ 0 ];
```
I don't think this is going to generate a unique name. Any given file could appear in more than one chunk. It's unlikely, but very possible.

blowery replied:
Also, given you're defaulting to `[]`, `name` could be undefined, which isn't a valid map key for the next line.

blowery replied:
Maybe instead of building the name from the filename, we could use the chunkhash?

yuliyan replied:
`chunk.files` contains the .js file name and in some cases its .css and .rtl.css that each chunk will output, not the individual modules it includes. So, it should be safe to use.

As for the `[]` fallback value - seems like I forgot to remove it after debugging something else. It's not really needed, since each chunk has atleast its .js files.

blowery commented on client/webpack/generate-chunks-map-plugin.js:
```diff
@@ -0,0 +1,39 @@
+				const name = files.find( file => /\.js$/.test( file ) ) || files[ 0 ];
+
+				map[ name ] = [ ...chunk.modulesIterable ]
+					.map( ( { userRequest } ) => userRequest && path.relative( '.', userRequest ) )
```
can we do the map / filter work prior to the assignment? I find this statement odd to read. It looks like a plain assignment that's missing a `;` at first glance.

yuliyan replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/39873/commits/256ed82550d73b1e36e793120d6e48172e0517dd

akirk commented:
> I'm curious about how it deals with the contents of a chunk changing over time. Every release could change the chunk makeup.

We generate a whitelist of chunks (`translated-chunks.json`) that are available. If the chunks change, the translations won't be loaded because it's not on the whitelist (and thus show English). The changed chunks will be communicated to `i18n-deploy` using the `chunks-map.json` through which `i18n-deploy` can generate the new chunks. We could create a feedback mechanism that automatically re-deploys if the chunks change.

yuliyan committed:
Update code formatting

yuliyan committed:
Add download languages script

yuliyan committed:
Temporary disable translation chunks fetching and add use local language files

yuliyan committed:
Ignore chunks-map.json

yuliyan committed:
Generate language chunks WIP

yuliyan committed:
Add translation chunks flag

yuliyan committed:
Rename download languages bin

akirk commented:
Played around with it and it works really nicely (I had to remember to use `?flags=use-translation-chunks`). All strings load translated fine and it's nice to see the translatiion chunks being loaded alongside. Great work!

yuliyan committed:
Rename translated-chunks.json

Rename translated-chunks.json to language-manifest.json and include
locale along with translated chunks keys.

yuliyan committed:
Use all translation contexts

yuliyan committed:
Add download languages revisions

yuliyan committed:
Add better language chunks progress logging

yuliyan committed:
Remove chunks-map CircleCI job

yuliyan committed:
Fix stdout compatibility

yuliyan committed:
Add chunks map, translate and build language chunks scripts to Dockerfile

yuliyan committed:
Fix using accurate chunk id for fetching

yuliyan committed:
Add wpcom svn revision when building docker image

yuliyan committed:
Use plain JS for injected webpack code

yuliyan committed:
Execute translate script prior to build-languages

yuliyan committed:
Build translation chunks for all browserlist env simultaneously

yuliyan committed:
Fix installed chunks ids

yuliyan committed:
Remove ES6 syntax from injected webpack code

yuliyan committed:
Add installed chunks to page bootstrap

yuliyan committed:
Combine chunks and entrypoint js files for installedChunks

yuliyan committed:
Expand concatenated modules in chunks map

yuliyan committed:
Use webpack's public path to fetch language chunks

yuliyan committed:
Use local lang-revisions file

yuliyan committed:
Use revision as cache buster for translation chunks

yuliyan committed:
Revert use internal translations feature flag

lynnbgriffith reviewed (approved):
👍🏻 Overall, this tests really well for me locally.  

Do you have any thoughts about this from a developer experience standpoint?  If I want to test i18n locally, currently I implicitly rely on the latest `i18n-deploy` artifacts and I can see translations after just a normal `npm start`.  Now, I need to know to use `CHUNKS_MAP=true` and to run `npm run build-languages`.
If I skip `CHUNKS_MAP`, I'll be hitting a code path that differs from production.  If I skip `build-languages`, I'll get no translations.  
I realize some of this will sort itself out when we're no longer gated by the `use-translation-chunks` feature flag, but I want to ensure we end up with a process for local testing that is as straightforward as possible.

yuliyan committed:
Remove unnecessary chunk files fallback

yuliyan committed:
Remove language dirs in favour of prefixing files

yuliyan committed:
Simplify translation chunks build flow

yuliyan commented on client/boot/locale.js:
```diff
@@ -23,5 +25,71 @@ export const setupLocale = ( currentUser, reduxStore ) => {
+		const getTranslationChunkPath = ( chunkId, langSlug = currentUser.localeSlug ) => {
+			const languageRevision = languageRevisions[ langSlug ] || '';
+
+			return `${ languagesPath }/${ langSlug }/${ chunkId }.json?${ languageRevision }`;
```
I've added the language slug as a prefix for the chunks and it looks a bit cleaner in the dev tools now.

yuliyan commented:
> 👍🏻 Overall, this tests really well for me locally.
> 
> Do you have any thoughts about this from a developer experience standpoint? If I want to test i18n locally, currently I implicitly rely on the latest `i18n-deploy` artifacts and I can see translations after just a normal `npm start`. Now, I need to know to use `CHUNKS_MAP=true` and to run `npm run build-languages`.
> If I skip `CHUNKS_MAP`, I'll be hitting a code path that differs from production. If I skip `build-languages`, I'll get no translations.
> I realize some of this will sort itself out when we're no longer gated by the `use-translation-chunks` feature flag, but I want to ensure we end up with a process for local testing that is as straightforward as possible.

Thanks for the code review!

I've reworked translation chunks build flow a bit, so it will now only require to pass `ENABLE_FEATURES=use-translation-chunks` when booting Calypso locally, e.g. `ENABLE_FEATURES=use-translation-chunks npm start`. It does the same as before, but won't require the extra steps. See updated testing instructions.

As for docker build, I've added a separate flag `BUILD_TRANSLATION_CHUNKS` that will still build translation chunks, but won't enable the feature by default, so it will still require adding `use-translation-chunks` flag to the URL.

akirk reviewed (approved):
I've tested and it works really well, I just noticed that the chunks are only built (and language files downloaded) with `ENABLE_FEATURES=use-translation-chunks npm start` when a `chunks-map.evergreen.json` already exists. Therefore this only works on a second run. I've talked to @yuliyan and he'll fix that in a follow-up. This has become big and merging this will allow us to test performance differences in production using the URL flag. Thanks for your great work, Yuliyan! Let's 🚢
